geps/gep-2198.md

```diff
@@ -0,0 +1,103 @@
+
+Here are some samples of JWT implementations in a similar problem space.  Each implementation differs in terminology and implementation.
+
+- NGINX Ingress Controller:
+
+  - [](https://docs.nginx.com/nginx-ingress-controller/configuration/policy-resource/#jwt-using-local-kubernetes-secret)
+  - [](https://docs.nginx.com/nginx-ingress-controller/configuration/policy-resource/#jwt-using-jwks-from-remote-location)
+  
+- Kong:
+
+  - [https://docs.konghq.com/hub/kong-inc/jwt/configuration/](https://docs.konghq.com/hub/kong-inc/jwt/configuration/)
+
+- Envoy Gateway:
+
+  - [https://gateway.envoyproxy.io/v0.4.0/api/extension_types.html#jwtauthenticationfilterprovider](https://gateway.envoyproxy.io/v0.4.0/api/extension_types.html#jwtauthenticationfilterprovider)
+
+- Istio:
+
+  - [https://istio.io/latest/docs/tasks/security/authorization/authz-jwt/](https://istio.io/latest/docs/tasks/security/authorization/authz-jwt/)
+
+## Properties of a JWT Validation Policy
+
+The validation policy is two fold, it establishes the relationship with the Token and how to validate it.
+According to [https://datatracker.ietf.org/doc/html/rfc7519#section-4.1](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1) claims cannot be assumed mandatory.  At a minimum a claim should be able to be validated following the process outlined here which only requires a method to decrypt if necessary: [https://datatracker.ietf.org/doc/html/rfc7519#section-7.2](https://datatracker.ietf.org/doc/html/rfc7519#section-7.2)
+
+The statements made above point to the primary points of divergence between the JWT validation implementations in the prior art section.
+
+Some very high level proposed properties to open the conversation based on comparing the implementations and reviewing the RFC.
+
+- name / realm (required) - the name / realm of the JWT, this is commonly a string representing a specific application.  It is seen used in response and log messages.
+- secret (optional) - The name of the Kubernetes secret that stores the JWK.
+- jwksURI (optional) - The remote URI where the request will be sent to retrieve JSON Web Key set
+- keyCache (extension) - Enables in-memory caching of JWKS (JSON Web Key Sets) that are obtained from the jwksURI and sets a valid time for expiration.
+- token (extension) - The token specifies a variable that contains the JSON Web Token in the request. By default the JWT is passed in the Authorization header as a Bearer Token. JWT may be also passed as a cookie or a part of a query string, for example: $cookie_auth_token. Accepted variables are $http_, $arg_, $cookie_.
+
+Example patterns:
+name
+secret
+
+name
+jwksURI
+
+name
+token
+jwksURI
+keycache
+
+## Properties of a JWT Claims (authorization) Policy
+
+- token - the token that needs to be evaluated (this is the result of the validation Policy or an OIDC workflow)
+- claimsMatch - an array of claim properties (exists, not exists) or match value that must match to allow the request to proceed.
```

Comment:
In addition to allowing the request to proceed, Istio provides access to the evaluated token in route rules, so that users can also do things like weighted routing based on the claims. See https://istio.io/latest/docs/tasks/security/authentication/jwt-route/

Reply:
I have historically considered this a claims match modifier (Policy) on a path/route.
First the path/route must match, then the claim is tested for allow/deny
I am assuming the same implementation.
This is where I see the authorization Policy serving the function of defining the allow claim(s) match.

Reply:
I do like the pattern with:
`    outputClaimToHeaders:
    - header: "x-jwt-claim-foo"
      claim: "foo"`
What I am thinking about is if there is a way to have the independent of the jwksUri, so that it stands alone as a transformation action pattern that could be used for both OIDC and JWT.  Since both can require the same claim-to-header, but both don't require the jwksUri.
Or the pattern is simply repeated when the OIDC Policy comes...